### ISSUE TITLE
Table: Better table widths

### ISSUE THREAD

**Issue by kdheepak** [ratatui-org/ratatui#537]
## Problem

Right now there's no easy way to apply a custom width for each columns. I have to resort to using `.widths(&[Constraint::XXX(), ...])`. 

For example, I want to be able to calculate a layout and pass in widths for the table to use, so that I can add a keybinding for the user to increase the width of a particular column. The way to do it at the moment is do a layout calculation, get the widths, then again explicitly call `.widths(widths.iter().map(|w| Constraint::Length(w)).collect::Vec<Constraint>())` to generate the specific widths that I want.

It would be nice if we could rename the current `.widths` to `.width_constraints()` and have a new `.widths()` function that takes just a `width` directly. This would have to happen over multiple versions so that it can deprecate the old name, and replace it with a new name. Alternatively we can use a different name instead.

See https://github.com/ratatui-org/ratatui/issues/510#issuecomment-1735116842 and https://github.com/ratatui-org/ratatui/issues/510#issuecomment-1735156677 for more context

**Comment by kdheepak** [comment]
Another issue at the moment is there's no way to override [`segment_size(SegmentSize::None)`](https://github.com/ratatui-org/ratatui/blob/301366c4fa33524b0634bbd3dcf1abd1a1ebe7c6/src/widgets/table.rs#L414). This makes it difficult to generate a table that fills available width. 

This would be partially addressed by https://github.com/ratatui-org/ratatui/issues/536 because if that's public, we can generate a custom layout with the intended widths and pass `Length` constraints to the table widget.

**Comment by joshka** [comment]
The simple non breaking approach to this might be to add:
```rust
impl Constraint {
  fn from_lengths(&[u16]) -> Vec<Self> {
    arr.iter().map(|&i| Self { value: i }).collect()
  }
}
```

Another thing that's worth noting about the table is the logic for ensuring that the gaps between columns get created.
We might need to customize the table logic a bit more to choose which constraints to ignore if there's not enough space to render the widths in the available space. E.g.:
1. Columns should render the gap between them
2. Collapse two adjacent gaps into a single gap
3. Start truncating from the right side when things don't fit to ensure things fit in the viewport
4. ?

**Comment by joshka** [comment]
> The simple non breaking approach to this might be to add:

https://github.com/ratatui-org/ratatui/pull/641

**Comment by asomers** [comment]
> Another issue at the moment is there's no way to override [`segment_size(SegmentSize::None)`](https://github.com/ratatui-org/ratatui/blob/301366c4fa33524b0634bbd3dcf1abd1a1ebe7c6/src/widgets/table.rs#L414). This makes it difficult to generate a table that fills available width.
> 
> This would be partially addressed by #536 because if that's public, we can generate a custom layout with the intended widths and pass `Length` constraints to the table widget.

How does one pass a `Layout` to a new `Table`?  I  can see how to set the `SegmentSize` for a `Layout`, but not how to communicate that to a `Table`.  Is that functionality missing, even after PR 536 merged?

**Comment by joshka** [comment]
> How does one pass a `Layout` to a new `Table`? I can see how to set the `SegmentSize` for a `Layout`, but not how to communicate that to a `Table`. Is that functionality missing, even after PR 536 merged?

Fixed in your PR :) https://github.com/ratatui-org/ratatui/pull/660
